File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int m = matrix.size();
        int n = matrix[0].size();
        
        int si = 0;
        int ei = (n*m)-1;
        
        while(si <= ei){
            int mid = si + (ei-si)/2;
            if(matrix[mid/n][mid%n] == target) return true;
            else if(matrix[mid/n][mid%n] < target) si = mid+1;
            else ei = mid-1;
            
            //cout<<matrix[mid][0]<<" ";
        }
        
//         int ri = si-1;
        
//         si = 0;
//         ei = n-1;
//         cout<<endl<<"ri "<<ri<<endl;
        
//         while(si <= ei){
//             int mid = si + (ei-si)/2;
            
//             if(matrix[ri][mid] == target) return true;
//             else if(matrix[ri][mid] < target) si = mid+1;
//             else ei = mid-1;
//             cout<<matrix[ri][mid]<<" ";
//         }
        
        return false;
        
    }
};

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        topR, bottomR = 0, len(matrix)-1
        
        while topR<=bottomR:
            mid = (topR+bottomR)//2
            
            if matrix[mid][0] <= target:
                topR=mid+1
            else:
                bottomR = mid-1
        
        left, right = 0, len(matrix[0])-1
        
        while left<=right:
            mid = (left+right)//2
            
            if matrix[bottomR][mid] == target:
                return True
            elif matrix[bottomR][mid] < target:
                left = mid+1
            else:
                right = mid-1
        
        
        return False